examples/go/charm/charmcraft.yaml

```diff
@@ -0,0 +1,76 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+name: go-k8s
+
+base: ubuntu@24.04
+
+platforms:
+  amd64:
+
+summary: An example Go application.
+
+description: An example Go application.
+
+actions:
+  rotate-secret-key:
+    description: Rotate the secret key. Users will be forced to log in again.
+      This might be useful if a security breach occurs.
+
+assumes:
+  - k8s-api
+
+config:
+  options:
+    port:
+      type: int
+      description: Default port where the application will listen
+      default: 8080
+    metrics-port:
+      type: int
+      description: Port where the prometheus metrics will be scraped.
+      default: 8080
+    metrics-path:
+      type: string
+      description: Path where the prometheus metrics will be scraped.
+      default: /metrics
+    secret-key:
+      type: string
+      description: Long secret you can use for sessions, csrf or any other thing where
+         you need a random secret shared by all units
+containers:
+  app:
+    resource: app-image
+peers:
+  secret-storage:
+    interface: secret-storage
+provides:
+  grafana-dashboard:
+    interface: grafana_dashboard
+  metrics-endpoint:
+    interface: prometheus_scrape
+requires:
+  ingress:
+    interface: ingress
+    limit: 1
+  logging:
+    interface: loki_push_api
+  postgresql:
+    interface: postgresql_client
+    optional: True
+    limit: 1
+resources:
+  app-image:
+    description: go application image.
+    type: oci-image
+type: charm
+
+parts:
+  charm:
+    charm-strict-dependencies: false
+    plugin: charm
+    build-snaps:
+    - rustup
+    override-build: |-
+      rustup default stable
+      craftctl default
```

Comment:
The extension in `charmcraft` already has this code, however in the examples the `charmcraft` extension is not used.

Reply:
How does a charmcraft extension work exactly? Does it only get invoked when the charm is first generated? Or does it somehow apply to charms with a certain property declared?

I see the [gunicorn extension](https://github.com/canonical/charmcraft/blob/4cce204c975df1e5d3c4f8ee73297bbffe860f21/charmcraft/extensions/gunicorn.py), but I don't quite grasp how it works in conjunction with paas-app-charmer.

Reply:
We use two things in charmcraft:
- Init profiles. They are created on "charmcraft init --profile go-framework". The go init profile uses these [template files](https://github.com/javierdelapuente/charmcraft/tree/go-extension/charmcraft/templates/init-go-framework)
- The extension. When the charmcraft.yaml file has a key "extension, like in "extension: [go-framework]", the extension runs with every invocation of charmcraft (for example in `charmcraft pack`). The extension "modifies" the yaml file before any of the parts lifecycle run. You can check the expanded yaml with `charmcraft expand-extensions`.

In our case, the extension adds all the required libs to the charmcraft.yaml (so charmcraft fetch-libs can be run), adds the action rotate-secret-key (and add user in django), interfaces, configuration options, resources, some checks in the yaml file...

Reply:
Got it, thanks! So since the extension is not used here, you have to manually include these details. Out of curiosity, why isn't the extension used in this example?

Reply:
The examples are used for unit tests too (see [here](https://github.com/canonical/paas-app-charmer/blob/c92936da305b411ff8e76f3ece545674fa96703b/tests/unit/go/conftest.py#L22)) and the ops testing  framework does not "expand" the charmcraft.yaml file. 

Two versions would be necessary and also building charmcraft against the PR, as it is not merged yet.